dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/DeserializerFactory.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.annotations;
+
+import com.palantir.dialogue.Deserializer;
+import com.palantir.dialogue.TypeMarker;
+
+public interface DeserializerFactory {
+    <T> Deserializer<T> deserializerFor(TypeMarker<T> type);
+}
```

Comment:
I think we may want to bound the factory to reduce ambiguity for deserializers that apply to a single type, while allowing the default json deserializer to apply to Object:

```java
public interface DeserializerFactory<T> {
    <U extends T> Deserializer<U> deserializerFor(TypeMarker<U> type);
}
```
I think the same applies to serializers.